stan/math/rev/functor/kinsol_data.hpp

```diff
@@ -71,13 +71,11 @@ class kinsol_system_data {
 
     Eigen::Map<Eigen::VectorXd> f_eval_map(N_VGetArrayPointer(f_eval),
                                            explicit_system->N_);
-
-    f_eval_map = apply(
-        [&](const auto&... args) {
-          return explicit_system->f_(x_eigen, explicit_system->msgs_, args...);
-        },
-        explicit_system->args_tuple_);
-
+    auto result = apply([&](const auto&... args) {
+      return explicit_system->f_(x_eigen, explicit_system->msgs_, args...);
+    }, explicit_system->args_tuple_);
+    check_matching_sizes("", "the algebraic system's output", result, "the vector of unknowns, x,", f_eval_map);
+    f_eval_map = result;
```

Comment:
This is the line that was failing the test, which it should have failed but was just failing badly. Should we be checking this at a higher level somewhere?

Reply:
By "checking this at a higher level," do you mean writing a more general test, or having the check occur somewhere else besides in the bodies of the `algebra_solver_*` functions?